src/analysis/processing/qgsalgorithmhubdistance.cpp

```diff
@@ -0,0 +1,253 @@
+      break;
+    case 2:
+      unit = Qgis::DistanceUnit::Miles;
+      break;
+    case 3:
+      unit = Qgis::DistanceUnit::Kilometers;
+      break;
+  }
+
+  QgsFields fields = spokeSource->fields();
+  fields.append( QgsField( QStringLiteral( "HubName" ), QMetaType::Type::QString ) );
+  fields.append( QgsField( QStringLiteral( "HubDist" ), QMetaType::Type::Double ) );
+
+  QString linesDest;
+  std::unique_ptr<QgsFeatureSink> linesSink( parameterAsSink( parameters, QStringLiteral( "OUTPUT_LINES" ), context, linesDest, fields, Qgis::WkbType::LineString, hubSource->sourceCrs(), QgsFeatureSink::RegeneratePrimaryKey ) );
+  if ( !linesSink )
+    throw QgsProcessingException( invalidSinkError( parameters, QStringLiteral( "OUTPUT_LINES" ) ) );
+
+  QString pointsDest;
+  std::unique_ptr<QgsFeatureSink> pointsSink( parameterAsSink( parameters, QStringLiteral( "OUTPUT_POINTS" ), context, pointsDest, fields, Qgis::WkbType::Point, hubSource->sourceCrs(), QgsFeatureSink::RegeneratePrimaryKey ) );
+  if ( !pointsSink )
+    throw QgsProcessingException( invalidSinkError( parameters, QStringLiteral( "OUTPUT_POINTS" ) ) );
+
+  QgsFeatureRequest request;
+  request.setSubsetOfAttributes( QgsAttributeList() << hubNameIndex );
+  request.setDestinationCrs( spokeSource->sourceCrs(), context.transformContext() );
+
+  const QgsSpatialIndex hubsIndex( hubSource->getFeatures( request ), feedback, QgsSpatialIndex::FlagStoreFeatureGeometries );
+  QHash<QgsFeatureId, QVariant> hubsAttributeCache;
+  double step = hubSource->featureCount() > 0 ? 50.0 / hubSource->featureCount() : 1;
+  long long i = 0;
+  const QgsSpatialIndex idx( hubSource->getFeatures( request ), [&]( const QgsFeature &f ) -> bool {
+    if ( feedback-> isCanceled() )
+    {
+      return false;
+    }
+
+    hubsAttributeCache.insert( f.id(), f.attributes().at( hubNameIndex ) );
+
+    i++;
+    feedback->setProgress( i * step );
+    return true; }, QgsSpatialIndex::FlagStoreFeatureGeometries );
+
+  QgsDistanceArea da;
+  da.setSourceCrs( spokeSource->sourceCrs(), context.transformContext() );
+  da.setEllipsoid( context.ellipsoid() );
+
+  // Scan source points, find nearest hub, and write to output file
+  i = 0;
+  QgsFeature spokeFeature, outputFeature;
+  step = spokeSource->featureCount() > 0 ? 50.0 / spokeSource->featureCount() : 1;
+  QgsFeatureIterator features = spokeSource->getFeatures();
+  while ( features.nextFeature( spokeFeature ) )
+  {
+    if ( feedback->isCanceled() )
+    {
+      break;
+    }
+
+    i++;
+    feedback->setProgress( i * step );
+
+    if ( !spokeFeature.hasGeometry() )
+    {
+      if ( linesSink && !linesSink->addFeature( spokeFeature, QgsFeatureSink::Flag::FastInsert ) )
+      {
+        throw QgsProcessingException( writeFeatureError( linesSink.get(), parameters, QStringLiteral( "OUTPUT_LINES" ) ) );
+      }
+      if ( pointsSink && !pointsSink->addFeature( spokeFeature, QgsFeatureSink::Flag::FastInsert ) )
```

Comment:
i think we need to pad the attributes here with two nulls, to account for the HubName/HubDist fields